src/gbinder_servicemanager.c

```diff
@@ -433,6 +504,81 @@ gbinder_servicemanager_cancel(
     }
 }
 
+gulong
+gbinder_servicemanager_add_registration_handler(
+    GBinderServiceManager* self,
+    const char* name,
+    GBinderServiceManagerRegistrationFunc func,
+    void* user_data) /* Since 1.0.13 */
+{
+    gulong id = 0;
+
+    if (G_LIKELY(self) && G_LIKELY(func) && name && name[0]) {
+        GBinderServiceManagerClass* klass =
+            GBINDER_SERVICEMANAGER_GET_CLASS(self);
+        GBinderServiceManagerPriv* priv = self->priv;
+        GBinderServiceManagerWatch* watch = NULL;
+        char* sig;
+
+        watch = g_hash_table_lookup(priv->watch_table, name);
+        if (!watch) {
+            watch = gbinder_servicemanager_watch_new(name);
+            g_hash_table_insert(priv->watch_table, watch->name, watch);
+        }
+        if (!watch->watched) {
+            watch->watched = klass->watch(self, name);
+            if (watch->watched) {
+                GDEBUG("Watching %s", watch->name);
+            } else {
+                GWARN("Failed to watch %s", watch->name);
+            }
+        }
+
+        sig = g_strconcat(SIGNAL_REGISTRATION_NAME "::", watch->detail, NULL);
+        id = g_signal_connect(self, sig, G_CALLBACK(func), user_data);
+        g_free(sig);
+    }
+    return id;
+}
+
+void
+gbinder_servicemanager_remove_handler(
+    GBinderServiceManager* self,
+    gulong id) /* Since 1.0.13 */
+{
+    if (G_LIKELY(self) && G_LIKELY(id)) {
+        GBinderServiceManagerClass* klass =
+            GBINDER_SERVICEMANAGER_GET_CLASS(self);
+        GBinderServiceManagerPriv* priv = self->priv;
+        GHashTableIter it;
+        gpointer value;
+
+        g_signal_handler_disconnect(self, id);
+        g_hash_table_iter_init(&it, priv->watch_table);
+        while (g_hash_table_iter_next(&it, NULL, &value)) {
+            GBinderServiceManagerWatch* watch = value;
+
+            if (g_signal_has_handler_pending(self,
+                gbinder_servicemanager_signals[SIGNAL_REGISTRATION],
+                watch->quark, TRUE)) {
+                if (!watch->watched) {
+                    GDEBUG("Watching %s", watch->name);
+                    watch->watched = klass->watch(self, watch->name);
+                    if (watch->watched) {
```

Comment:
Does this ever happend? Unless g_signal_connect is called manually elsewhere.

Reply:
Right, very unlikely scenario. I wasn't sure if it needs to be handled at all but the expensive part here is `g_signal_has_handler_pending()` which is needed anyway since we don't know which one was removed; so I thought why not to do another check for free...

On the other hand, if API is not used properly and `g_signal_connect()`/`g_signal_handler_disconnect()` are used directly, the caller is screwed anyway and there's no need to handle scenarios like this one. The code can be even slightly optimized - as soon as we find a watched name that has no handlers, we can unwatch it and break the loop, assuming that we have found the name watched by the handler that we have just removed (and it was the last one for this name).

In reality it all doesn't really matter because the most common use case would be a single listener for the entire process. There's very little to to optimize here.

Reply:
Agreed, no real need for optimization. With custom closure type, g_signal_handler_disconnect could actually be made to work (and allow implementing something like _add_handler_object).

Reply:
Meh, GBinderServiceManager is not even declared in the header file as a GObject. The caller should not assume that it is.

I simplified this piece of code a bit, dropping the check for improper use of API which doesn't really guarantee anything.